### PR TITLE
Merge desc siol

### PR DESCRIPTION
**Pull request by condret**
Yes, it causes breakage in the testsuite. But fixing this, would be easier on master, for social reasons. 3 very busy ppl cannot do this alone. So, hopefully merging this into master causes more devs to fix it. SIOL must be done, or esil won't emulate most programs correctly. Trust me, I'm a reverse-engineer!

### CODE REVIEW

File: libr/io/io.c [hunk 1]
```diff
@@ -769,28 +666,53 @@ R_API int r_io_write(RIO *io, const ut8 *buf, int len) {
...
+    }
+    if (io->cached) {
+        ret = r_io_cache_write (io, addr, mybuf, len);
+    } else if (io->va) {
+        ret = r_io_vwrite_at (io, addr, mybuf, len);
+    } else {
+        ret = r_io_pwrite_at (io, addr, mybuf, len);
+    }
+    if (buf != mybuf) {
+        free (mybuf);
```
Comment (radare) [inline libr/io/io.c]: Use tabs

File: libr/io/io.c [hunk 2]
```diff
@@ -889,25 +794,7 @@ R_API bool r_io_is_blockdevice(RIO *io) {
 }
 
 R_API ut64 r_io_size(RIO *io) {
```
Comment (radare) [inline libr/io/io.c]: Imho iosize should be removed or return the total size of memory allocated. Not just the current file

File: libr/io/io.c [hunk 3]
```diff
@@ -889,25 +794,7 @@ R_API bool r_io_is_blockdevice(RIO *io) {
...
-		return UT64_MAX;
-	}
-	io->va = false;
-	here = r_io_seek (io, 0, R_IO_SEEK_CUR);
-	size = r_io_seek (io, 0, R_IO_SEEK_END);
-	if (r_io_seek (io, here, R_IO_SEEK_SET) != here) {
-		eprintf ("Failed to reset the file position\n");
-	}
-	io->va = oldva;
-	if (r_io_is_blockdevice (io)) {
```
Comment (radare) [inline libr/io/io.c]: Without this opening a block device will fail because it will consider the size of 0 and not read beyond

Comment (radare) [inline libr/io/io.c]: Did u tested this?

File: libr/io/p/io_debug.c [hunk 4]
```diff
@@ -1,4 +1,4 @@
-/* radare - LGPL - Copyright 2007-2017 - pancake */
+/* radare - LGPL - Copyright 2007-2016 - pancake */
```
Comment (radare) [inline libr/io/p/io_debug.c]: Its 2017

File: libr/io/p/io_debug.c [hunk 5]
```diff
@@ -68,31 +68,31 @@ static void inferior_abort_handler(int pid) {
...
-        if (tok) {
+	if (tok) {
 		CloseHandle (tok);
 	}
-        if (err) {
+	if (err) {
 		r_sys_perror ("setup_tokens");
 	}
-        return err;
+	return err;
```
Comment (radare) [inline libr/io/p/io_debug.c]: Tabs

File: libr/io/p/io_debug.c [hunk 6]
```diff
@@ -106,7 +106,7 @@ static int fork_and_ptraceme(RIO *io, int bits, const char *cmd) {
 	}
 	setup_tokens ();
 	char *_cmd = io->args ? r_str_appendf (strdup (cmd), " %s", io->args) :
-				strdup (cmd);
+	strdup (cmd);
```
Comment (radare) [inline libr/io/p/io_debug.c]: Strange ternary operator. Start lines with ? And : for readabikity

File: libr/io/p/io_debug.c [hunk 7]
```diff
@@ -180,27 +180,20 @@ static int fork_and_ptraceme(RIO *io, int bits, const char *cmd) {
...
 static void trace_me () {
-#if __APPLE__
-	signal (SIGTRAP, SIG_IGN); //NEED BY STEP
-#endif
-#if __APPLE__ || __BSD__
-/* we can probably remove this #if..as long as PT_TRACE_ME is redefined for OSX in r_debug.h */
-	signal (SIGABRT, inferior_abort_handler);
-	if (ptrace (PT_TRACE_ME, 0, 0, 0) != 0) {
-#else
 	if (ptrace (PTRACE_TRACEME, 0, NULL, NULL) != 0) {
```
Comment (radare) [inline libr/io/p/io_debug.c]: This will not build on bsd i think

File: libr/io/p/io_default.c [hunk 8]
```diff
@@ -240,31 +245,31 @@ static int r_io_def_mmap_write(RIO *io, RIODesc *fd, const ut8 *buf, int count)
...
 				return -1;
 			}
-			a_count = count + (aligned-(count%aligned));
-
-			a_buf = malloc (a_count+aligned);
+			a_count = count + (aligned - (count % aligned));
+			a_buf = malloc (a_count + aligned);
 			if (a_buf) {
 				int i;
 				memset (a_buf, 0xff, a_count+aligned);
```
Comment (radare) [inline libr/io/p/io_default.c]: Missing space around +

File: libr/io/p/io_gdb.c [hunk 9]
```diff
@@ -1,4 +1,4 @@
-/* radare - LGPL - Copyright 2010-2017 pancake */
+/* radare - LGPL - Copyright 2010-2016 pancake */
```
Comment (radare) [inline libr/io/p/io_gdb.c]: Its 2017

File: libr/io/p/io_default.c [hunk 10]
```diff
@@ -293,11 +298,13 @@ static int r_io_def_mmap_write(RIO *io, RIODesc *fd, const ut8 *buf, int count)
 
 static RIODesc *r_io_def_mmap_open(RIO *io, const char *file, int flags, int mode) {
 	RIOMMapFileObj *mmo = r_io_def_mmap_create_new_file (io, file, mode, flags);
-	return (mmo)
-		? r_io_desc_new (&r_io_plugin_default, mmo->fd, mmo->filename, flags, mode, mmo)
-		: NULL;
+	if (!mmo) {
```
Comment (radare) [inline libr/io/p/io_default.c]: The ternary operator was nicer imho

File: libr/io/p/io_gdb.c [hunk 11]
```diff
@@ -140,16 +142,8 @@ static RIODesc *__open(RIO *io, const char *file, int rw, int mode) {
 				eprintf ("gdbr: Failed to attach to PID %i\n", i_pid);
 				return NULL;
 			}
-		} else if ((i_pid = desc->pid) < 0) {
-			i_pid = -1;
-		}
-		// Get name
-		char *name = gdbr_exec_file_read (desc, i_pid);
```
Comment (radare) [inline libr/io/p/io_gdb.c]: Regression. This code shouldnt be removed

File: libr/io/p/io_gdb.c [hunk 12]
```diff
@@ -323,11 +315,3 @@ RIOPlugin r_io_plugin_gdb = {
 	.system = __system,
 	.isdbg = true
 };
-
-#ifndef CORELIB
-RLibStruct radare_plugin = {
```
Comment (radare) [inline libr/io/p/io_gdb.c]: Wtf

File: libr/io/p/io_gzip.c [hunk 13]
```diff
@@ -118,18 +117,17 @@ static bool __plugin_open(RIO *io, const char *pathname, bool many) {
...
-			mal->buf = r_inflate (data, len, NULL, &mal->size);
-			if (mal->buf) {
-				RETURN_IO_DESC_NEW (&r_io_plugin_malloc,
-					mal->fd, pathname, rw, mode, mal);
-			}
-			free (data);
-			eprintf ("Cannot allocate (%s) %d bytes\n", pathname + 9, mal->size);
-			free (mal);
-		}
+		if (!mal) return NULL;
```
Comment (radare) [inline libr/io/p/io_gzip.c]: Braces

**Comment by MaskRay** [comment]
I can give a hand! Trust me, albeit I'm not a reverse engineer!

File: libr/io/p/io_gzip.c [hunk 14]
```diff
@@ -118,18 +117,17 @@ static bool __plugin_open(RIO *io, const char *pathname, bool many) {
...
+		int len;
+		ut8 *data = (ut8*)r_file_slurp (pathname+7, &len);
+		mal->buf = r_inflate (data, len, NULL, &mal->size);
+		if (mal->buf)
```
Comment (radare) [inline libr/io/p/io_gzip.c]: Braces

File: libr/io/p/io_ihex.c [hunk 15]
```diff
@@ -382,8 +381,7 @@ static RIODesc *__open(RIO *io, const char *pathname, int rw, int mode) {
 			free (str);
 			return NULL;
 		}
-		mal->fd = -1; /* causes r_io_desc_new() to set the correct fd */
-		mal->rbuf = r_buf_new_sparse ();
+		mal->rbuf = r_buf_new_sparse();
```
Comment (radare) [inline libr/io/p/io_ihex.c]: Missing space before (

File: libr/io/p/io_mach.c [hunk 16]
```diff
@@ -457,6 +471,14 @@ static int __system(RIO *io, RIODesc *fd, const char *cmd) {
 	return 1;
 }
 
+static int __get_pid (RIODesc *desc) {
+	RIOMach *mach = desc ? (RIOMach *) desc->data : NULL;
+	if (mach) {
```
Comment (radare) [inline libr/io/p/io_mach.c]: Use ternary

File: libr/anal/p/anal_ws.c [hunk 17]
```diff
@@ -11,11 +11,11 @@
...
 
 static ut64 ws_find_label(int l, RIOBind iob) {
-	ut64 cur = 0, size = iob.size(iob.io);
+	RIO *io = iob.get_io (&iob);
+	ut64 cur = 0, size = iob.desc_size (io->desc);
 	ut8 buf[128];
-	RAsmOp *aop;
-	aop = R_NEW0(RAsmOp);
-	iob.read_at(iob.io, cur, buf, 128);
+	RAsmOp *aop = R_NEW0 (RAsmOp);
```
Comment (radare) [inline libr/anal/p/anal_ws.c]: Nullchk

File: libr/anal/p/anal_ws.c [hunk 18]
```diff
@@ -11,11 +11,11 @@
...
+	iob.read_at (iob.io, cur, buf, 128);
 	while(cur <= size && wsdis(aop, buf, 128)) {
```
Comment (radare) [inline libr/anal/p/anal_ws.c]: Missing spaces before (

File: libr/core/cconfig.c [hunk 19]
```diff
@@ -2576,6 +2593,7 @@ R_API int r_core_config_init(RCore *core) {
 	SETI ("io.buffer.from", 0, "Lower address of buffered cache");
 	SETI ("io.buffer.to", 0, "Higher address of buffered cache");
 	SETCB ("io.cache", "false", &cb_iocache, "Enable cache for io changes");
+	SETCB("io.pcache", "false", &cb_iopcache, "io.cache for p-level");
```
Comment (radare) [inline libr/core/cconfig.c]: Add missing space

File: libr/io/p/io_ptrace.c [hunk 20]
```diff
@@ -192,53 +194,17 @@ static RIODesc *__open(RIO *io, const char *file, int rw, int mode) {
...
-			RIODesc *desc;
 			RIOPtrace *riop = R_NEW0 (RIOPtrace);
-			if (!riop) return NULL;
+			if (!riop) {
+				return NULL;
+			}
 			riop->pid = riop->tid = pid;
 			open_pidmem (riop);
-#if 1
-			{
```
Comment (radare) [inline libr/io/p/io_ptrace.c]: Is this because 2doo is not working ?

File: libr/io/p/io_r2k.c [hunk 21]
```diff
@@ -65,7 +65,7 @@ static int r2k__close(RIODesc *fd) {
 	}
 #elif defined (__linux__) && !defined (__GNU__)
 	if (fd) {
-		close (fd->fd);
+		close ((int)fd->data);
```
Comment (radare) [inline libr/io/p/io_r2k.c]: Missing size_t cast after (int)

File: libr/io/p/io_r2k.c [hunk 22]
```diff
@@ -117,7 +119,10 @@ static RIODesc *r2k__open(RIO *io, const char *pathname, int rw, int mode) {
 		r2k_struct.beid = 0;
 		r2k_struct.pid = 0;
 		r2k_struct.wp = 1;
-		return r_io_desc_new (&r_io_plugin_r2k, fd, pathname, rw, mode, NULL);
+		//return r_io_desc_new (&r_io_plugin_r2k, fd, pathname, rw, mode, NULL);
+		iodesc = r_io_desc_new (io, &r_io_plugin_r2k, pathname, rw, mode, NULL);
```
Comment (radare) [inline libr/io/p/io_r2k.c]: Check if null

File: libr/io/p/io_sparse.c [hunk 23]
```diff
@@ -90,10 +87,9 @@ static RIODesc *__open(RIO *io, const char *pathname, int rw, int mode) {
 				free (data);
 			}
 		}
-		if (mal->buf) {
-			RETURN_IO_DESC_NEW (&r_io_plugin_sparse,
-				mal->fd, pathname, rw, mode, mal);
-		}
+		if (mal->buf)
```
Comment (radare) [inline libr/io/p/io_sparse.c]: Braces

File: libr/io/plugin.c [hunk 24]
```diff
@@ -16,72 +18,72 @@ R_API bool r_io_plugin_add(RIO *io, RIOPlugin *plugin) {
...
 R_API RIOPlugin *r_io_plugin_resolve(RIO *io, const char *filename, bool many) {
-	RListIter *iter;
-	RIOPlugin *iop;
-	r_list_foreach (io->plugins, iter, iop) {
-		if (iop->check && iop->check (io, filename, many)) {
-			return iop;
+	SdbListIter *iter;
+	RIOPlugin *ret;
+	ls_foreach (io->plugins, iter, ret) {
+		if (!ret) {
```
Comment (radare) [inline libr/io/plugin.c]: Group those two conditionals into a single one with ||

File: binr/radare2/radare2.c [hunk 25]
```diff
@@ -1176,7 +1176,7 @@ int main(int argc, char **argv, char **envp) {
 #endif
 #endif
 	if (fullfile) {
-		r_core_block_size (&r, r_io_desc_size (r.io, r.file->desc));
+		r_core_block_size (&r, r_io_desc_size (r.file->desc));
 	}
```
Comment (condret) [inline binr/radare2/radare2.c]: yes, it is inside RIODesc, but that is not the point here. RIOPlugin is inside RIODesc, too. For calculating the size, you don't need the io, if you have the desc and the plugin. So clean alternative would be r_io_size (io, fd), and having desc_size for internal use mostly

File: libr/core/asm.c [hunk 26]
```diff
@@ -398,7 +398,18 @@ R_API RList *r_core_asm_bwdisassemble (RCore *core, ut64 addr, int n, int len) {
 	}
 
 	buf = (ut8 *)malloc (len);
-	if (!buf || r_io_read_at (core->io, addr - len / addrbytes, buf, len) != len) {
+	if (!hits || !buf) {
+		if (hits) {
+			r_list_free (hits);
```
Comment (condret) [inline libr/core/asm.c]: everybody hates rebases

File: libr/core/cmd_open.c [hunk 27]
```diff
@@ -425,6 +425,27 @@ R_API void r_core_file_reopen_debug (RCore *core, const char *args) {
...
+	RIODesc *desc = (RIODesc *)data;
+	RIOMap *map;
+	SdbListIter *iter;
+	p->cb_printf ("%2d %c %s : %s size=0x%"PFMT64x"\n", desc->fd, 
+			(desc->io && (desc->io->desc == desc)) ? '*' : '-',
+			desc->uri, r_str_rwx_i (desc->flags), r_io_desc_size (desc));
+	if (desc->io && desc->io->va && desc->io->maps) {
+		ls_foreach_prev (desc->io->maps, iter, map) {
+			if (map->fd == desc->fd) {
+				p->cb_printf ("\t+0x%"PFMT64x" 0x%"PFMT64x
```
Comment (condret) [inline libr/core/cmd_open.c]: wy?

File: libr/debug/debug.c [hunk 28]
```diff
@@ -1511,24 +1510,13 @@ R_API ut64 r_debug_get_baddr(RDebug *dbg, const char *file) {
 		// Tell gdb that we want baddr, not full mem map
 		dbg->iob.system(dbg->iob.io, "baddr");
 	}
-#if __WINDOWS__
```
Comment (condret) [inline libr/debug/debug.c]: @xarkes said he is refactoring the windbg, this code is a very good example of things that should live in a plugin, but doesn't.

File: libr/core/cconfig.c [hunk 29]
```diff
@@ -850,15 +850,16 @@ static int cb_cfgdebug(void *user, void *data) {
...
 #else
-			r_debug_select (core->dbg, core->file->desc->fd,
-					core->file->desc->fd);
+			int pid = r_io_desc_get_pid (core->io, core->file->desc->fd);
+			r_debug_select (core->dbg, pid, pid);
 #endif
 		}
 	} else {
-		if (core->dbg) r_debug_use (core->dbg, NULL);
+		if (core->dbg) {
```
Comment (condret) [inline libr/core/cconfig.c]: not sure, i'd like to wait for the refactor of windbg, before touching this. The if WINDOWS must die

**Comment by XVilka** [comment]
How is this related to https://github.com/radare/radare2/pull/7427 ? Old one can be closed then?

File: libr/bin/bin.c [hunk 30]
```diff
@@ -877,11 +877,8 @@ R_API int r_bin_load(RBin *bin, const char *file, ut64 baseaddr, ut64 loadaddr,
 		r_io_bind (io, &bin->iob);
 	}
 	bin->rawstr = rawstr;
-	if (fd == -1) {
-		desc = iob->desc_open (io, file, O_RDONLY, 0644);
-	} else {
-		desc = iob->desc_get_by_fd (io, fd);
-	}
+	//Use the current RIODesc otherwise r_io_map_select can swap them later on
```
Comment (condret) [inline libr/bin/bin.c]: why should fd become -1?

File: libr/io/desc.c [hunk 31]
```diff
@@ -1,262 +1,216 @@
...
-		}
-		if (max == -1 || s->to > max) {
-			max = s->to;
-		}
+R_API int r_io_desc_get_tid(RIO* io, int fd) {
+	RIODesc* desc;
+	if (!io || !io->files) {
+		return -2;              //-1 is reserved for plugin internal errors
 	}
-	mul = (max-min) / width;
```
Comment (condret) [inline libr/io/desc.c]: I think I put it into the core, maybe we lost it for now. Printing stuff is not a task for io.

File: libr/io/io.c (hunk 2 above)
Comment (condret) [inline libr/io/io.c]: agree

File: libr/io/p/io_default.c (hunk 10 above)
Comment (condret) [inline libr/io/p/io_default.c]: sometimes you really have a bad taste

File: libr/io/p/io_gdb.c (hunk 12 above)
Comment (condret) [inline libr/io/p/io_gdb.c]: ancient plugin-definition

File: libr/io/p/io_ptrace.c (hunk 20 above)
Comment (condret) [inline libr/io/p/io_ptrace.c]: I don't think so

File: libr/io/p/io_debug.c (hunk 5 above)
Comment (condret) [inline libr/io/p/io_debug.c]: it is a tab

File: libr/anal/p/anal_ws.c [hunk 32]
```diff
@@ -14,19 +14,17 @@ static ut64 ws_find_label(int l, RIOBind iob) {
 	RIO *io = iob.get_io (&iob);
 	ut64 cur = 0, size = iob.desc_size (io->desc);
 	ut8 buf[128];
-	RAsmOp *aop = R_NEW0 (RAsmOp);
+	RAsmOp aop;;
```
Comment (MaskRay) [inline libr/anal/p/anal_ws.c]: superfluous `;`

I like the commit description :)

beauty-fixes for pancake
--

Comment (condret) [inline libr/anal/p/anal_ws.c]: most it was "use tabs" or "missing space"

**Comment by radare** [comment]
conflict

File: libr/debug/debug.c (hunk 28 above)
Comment (radare) [inline libr/debug/debug.c]: agree should we just merge that and wait for wusers complains? 

@skuater

**Comment by condret** [comment]
merged

**Pull request closed without merging**